frontend/js/maze1.js

```diff
@@ -0,0 +1,297 @@
+
+        if (!td) {
+            return;
+        }
+
+        if (this._choosing.point) {
+            if (td === this._choosing.point) {
+                this._choosing.point = null;
+
+                td.dataset.info = '';
+                this._choosing.toggle(td);
+            }
+        } else if (!td.dataset.info) {
+            this._choosing.point = td;
+
+            td.dataset.info = this._choosing.info;
+            this._choosing.toggle(td);
+        }
+
+        switch (this._choosing.info) {
+            case 'S': {
+                this._startPoint = this._choosing.point;
+                break;
+            }
+            case 'E': {
+                this._endPoint = this._choosing.point;
+                break;
+            }
+        }
+    }
+
+//------------- private methods---------------
+    _renderMaze() {
+        let mazeHtml = '<table data-selector="map">';
+
+        for(let i = 0; i<this.size; i++) {
+            mazeHtml +='<tr>\n';
+            for (let j = 0; j<this.size; j++) {
+                mazeHtml+='<td></td>\n';
+            }
+            mazeHtml +='</tr>';
+        }
+
+        mazeHtml +=  '</table>';
+        this._el.insertAdjacentHTML('beforeEnd',mazeHtml);
+    }
+
+    _waveExpansionStep() {
+        let step = this._step;
+        let wasMarked = false;
+
+        for (let i = 0; i < this.size; i++) {
+            for (let j = 0; j < this.size; j++) {
+                let td = this._map.rows[i].cells[j];
+
+                if (parseInt(td.innerHTML) === (step - 1)) {
+
+                    let neighbours = this.cellNeighbours(i,j);
+
+                    neighbours.filter(cell => !cell.innerHTML && !cell.dataset.info)
+                        .forEach(cell => {
+                            cell.innerHTML = step;
+                            wasMarked = true;
+                        });
+                }
+            }
+        }
+        this._step++;
+
+        return wasMarked;
+    }
+
+    _backTraceStart() {
+        let event = new CustomEvent("backTraceStart");
+
+        this._el.dispatchEvent(event);
+    }
+
+    _backTraceStep() {
+        if (!this._currentCell) {
+            this._currentCell = this.endPoint;
+        }
+
+        let neighbours = this.cellNeighbours(this._currentCell.parentNode.rowIndex, this._currentCell.cellIndex);
+
+        this._currentCell = this._nextCellWithMinSteps(neighbours,this._currentCell);
+
+        if (this._currentCell === this.startPoint) {
+            return false;
+        } else {
+            Maze._markPath(this._currentCell);
+            return true;
+        }
+
+    }
+
+    _finish(message) {
+        let event = new CustomEvent("finish", {
+            detail: {message: message}
+        });
+
+        this._el.dispatchEvent(event);
+    }
+
+    _nextCellWithMinSteps(cells, currentCell) {
+        return cells.reduce( (current, cell)=> {
+            if ( parseInt(cell.innerHTML) < parseInt(current.innerHTML) ) {
```

Comment:
You don't need to use HTML to check the condition you can have array with data
and look in it